File: dweb/StructuredBlock.py
# encoding: utf-8
from json import loads
from Block import Block
from SmartDict import SmartDict
from Dweb import Dweb

#TODO-BACKPORT - review this file

class StructuredBlock(SmartDict):
    """
    Encapsulates an JSON Dict and stores / retrieves over transports.
    Note that data is data contained *in* the SB, while _data is data representation *of* the SB.
    Similarly url is a pointer to data contained *in* the SB, while _url is the the url of the data representing the SB.

    Certain sets of data fields have meaning in different applications
    { data: str; url: multiurl; links [ StructuredBlock* ] } refer to content
    _signatures are used for signing in subclasses of CommonList (e.g. MutableBlock, ACL)


    """

    # Uses SmartDict._data and SmartDict._data.setter from superclass
    # SmartDict.__init__(url=None, data=None) used, it will call @_data.setter here
    table = "sb"

    #---------------------------------------------------------------------------------------------------------
    #METHODS THAT EXTEND OR REPLACE THOSE ON SMARTDICT or TRANSPORTABLE
    #Files have either url or data or links field set,
    #---------------------------------------------------------------------------------------------------------

    def __init__(self, data=None, url=None, verbose=False, **options):
        if verbose: print "StructuredBlock.__init__",url, data[0:50] if data else "None", options
        super(StructuredBlock, self).__init__(data=data, url=url, verbose=verbose, **options)
        self._signatures = []   #TODO-BACKPORT Was Signatures([])

    def store(self, verbose=False, **options):
        """
        Store content if not already stored (note it must have been stored prior to signing)
        Store any signatures in the Transport layer
        """
        if not self._url:
            super(StructuredBlock, self).store(verbose=verbose, **options)    # Sets self._url   #TODO-EFFICIENCY DONT STORE IF NOT CHANGED
        for s in self._signatures:
            ss = s.copy()
            self.transport().add(url=self._url, date = ss.date,
                                     signature = ss.signature, signedby = ss.signedby, verbose=verbose, **options)
        return self # For chaining

    @property
    def links(self):
        return self.__dict__.get("links")

    @links.setter
    def links(self, value):
        """
        Convert returned list to array of StructuredBlocks

        :param value:  [ link as dict or StructuredBlock
        :return:
        """
        self.__dict__["links"] = [l if isinstance(l, StructuredBlock) else StructuredBlock(l) for l in value ]

    def dirty(self):
        super(StructuredBlock, self).dirty()
        from Signature import Signatures
        self._signatures = Signatures([]) # Cant be signed if changed from stored version

    #---------------------------------------------------------------------------------------------------------
    #METHODS TO DEAL WITH FILES, WHICH ARE STRUCTURED BLOCKS BUT DONT HAVE OWN TYPE AS INCLUDED BY OTHER THINGS
    #Files have either url or data or links field set,
    #---------------------------------------------------------------------------------------------------------

    def link(self, item):
        """
        Find a link of a SB either by integer or by name

        :param item: int or string
        :return:
        """
        if isinstance(item, int):
            return self.links[item]
        else:
            for sl in self.links:
                if sl.name == item:
                    return sl
        return None

    def content(self, verbose=False, **options):
        """
        Return the content of the SB.
        Note that this has to use the url or data fields to get the content held, rather than _url or _data which represents the SB itself.
        :return: content of block, fetching links (possibly recursively) if required
        """
        return (
            self.data or
            (self.url and Dweb.transport(self.url).rawfetch(url = self.url, verbose=verbose, **options)) or # Hash must point to raw data, not another SB
            (self.links and "".join(l.content(verbose=verbose, **options) for l in self.links)) or # Each link is a SB
            "")

    def file(self, contenttype=None, verbose=False, **options):
        """
        Return content as a dict, with any meta-data for HTTP to return

        :param contenttype:
        :param verbose:
        :param options:
        :return:
        """
        if verbose: print "StructuredBlock.file: contenttype=",contenttype,"options=",options,self
        self.fetch(verbose=verbose, **options)
        return {
            "Content-type": contenttype or self.__dict__.get("Content-type", None) or "application/octet-stream",
            "data": self.content(verbose=verbose, **options)
        }



    def size(self, verbose=False, **options):
        """
        Return the length of the content of the SB.
        Note that this has to use the url or data fields to get the content held, rather than _url or _data which represents the SB itself.

        :return: Size of content, or None if cant calculate
        """
        if verbose: print "SB.size: size=", self.__dict__.get("size",None), "data=", self.data, "url=", self.url, "links=", self.links
        return (
            self.__dict__.get("size",None) or
            (self.data and len(self.data)) or
            (self.url and len(Dweb.transport(self.url).rawfetch(url = self.url, verbose=verbose, **options))) or
            (self.links and sum(l.size(verbose=verbose, **options) for l in self.links)) or # Each link is a SB
            None)

    def path(self, urlargs, verbose=False, **optionsignored):
        """
        Walk a path and return the SB at the end of that path

        :param urlargs:
        :return:    Found path or None
        """
        self.fetch(verbose=verbose)
        sb = self
        while urlargs:
            sb = sb.link(urlargs.pop(0))
        return sb


    #---------------------------------------------------------------------------------------------------------
    #METHODS TO DEAL WITH SIGNATURES, STORED IN _signatures field
    #Files have either url or data or links field set,
    #---------------------------------------------------------------------------------------------------------

    def sign(self, commonlist, verbose=False, **options):
        """
        Add a signature to a StructuredBlock
        Note if the SB has a _acl field it will be encrypted first, then the url of the encrypted block used for signing.

        :param CommonList commonlist:   List its going on - has a ACL with a private key
        :return: self
        """
        from Signature import Signature
        if not self._url:
            self.store()  # Sets _url which is needed for signatures #TODO-EFFICIENCY only store if not stored
        self._signatures.append(Signature.sign(commonlist=commonlist, url=self._url, verbose=verbose))
        return self  # For chaining

    def verify(self, verbose=False, verify_atleastone=False, **options):
        """
        Verify the signatures on a block (if any)

        :param verbose: True for debugging output
        :param verify_atleastone: True if should fail if no signatures
        :param options: unused
        :return: True if all signatures present match
        """
        if verbose: print "StructuredBlock.verify", self
        if verify_atleastone and not len(self._signatures):
            return False
        return self._signatures.verify(url=self._url)

    #---------------------------------------------------------------------------------------------------------
    #METHODS TO DEAL WITH ENCRYPTION, STORED IN _acl field
    #Files have either url or data or links field set,
    #---------------------------------------------------------------------------------------------------------

    #Javascript has _date, earliestdate() & compare() to support dating based on signatures.

File: js/StructuredBlock.js
const SmartDict = require("./SmartDict");
const Signature = require("./Signature");
const Dweb = require("./Dweb");


// ######### Parallel development to StructuredBlock.py ########

class StructuredBlock extends SmartDict {

    //p_fetch uses SmartDict ..
    constructor(hash, data, verbose) {
        super(hash, data, verbose); // _hash is _hash of SB, not of data
        this._signatures = [];
        this._date = null;  // Updated in _earliestdate when loaded
        this.table = "sb";  // Note this is cls.table on python but need to separate from dictionary
    }
    p_store(verbose) {
        /*
         Store content if not already stored (note it must have been stored prior to signing)
         Store any signatures in the Transport layer
         Resolution of promise will happen on p_store, the addition of signatures will happen async - could change to Promise.all
         */
        if (!this._hash) {
            return super.p_store(verbose);    //Sets self._hash and stores in background if has changed
        }
        for (let i in this._signatures) {
            let s = this._signatures[i];
            //PY makes copy of s, but this is because the json procedure damages the object which doesnt happen in Crypto.dumps in JS
            Dweb.transport.p_add(this._hash, s.date, s.signature, s.signedby, null, verbose);
        }
        return this; // For chaining

    }

    __setattr__(name, value) {
        // Call chain is ...  or constructor > _setdata > _setproperties > __setattr__
        // catch equivalent of getters and setters here
        let verbose = false;
        if (name === "links") {  // Assume its a SB TODO make dependent on which table
            let links = value;
            for (let len = links.length, i=0; i<len; i++) {
                links[i] = new StructuredBlock(null, links[i], verbose);
            }
            this[name] = links;
        } else {
            super.__setattr__(name, value);
        }
    }

    p_path(patharr, verbose, successmethod) {
        // We cant use a function here, since the closure means it would apply to the object calling this, not the object loaded.
        // successmethod is an array [ nameofmethod, args... ]
        // Warning - may change the signature of this as discover how its used.
        if (verbose) { console.log("sb.p_path",patharr, successmethod, "links=",this.links); }
        if (patharr && patharr.length && this.links && this.links.length ) { // Have a path and can do next bit of it on this sb
            let next = patharr.shift(); // Takes first element of patharr, leaves patharr as rest
            if (verbose) { console.log("StructuredBlock:path next=",next); }
            let obj = this.link(next);   //TODO handle error if not found
            return obj.p_fetch(verbose).then(() => obj.p_path(patharr, verbose))
        } else if (patharr && patharr.length) {
            throw new Error("Cant follow path"+patharr);
        } else {  // Reached end of path, can apply success
            //TODO-IPFS unsure if have this correct
            //if (success) { success(undefined); } //note this wont work as success defined on different object as "this"
            if (successmethod) {
                let methodname = successmethod.shift();
                if (verbose) { console.log("p_path success:",methodname, successmethod); }
                this[methodname](...successmethod); // Spreads successmethod into args, like *args in python
            }
            return new Promise((resolve, reject)=> resolve(null));  // I think this should be a noop - fetched already
        }
    }

    dirty() {
        super.dirty();
        this._signatures = [];
    }

    link(name) {    // Note python version allows call by number as well as name
        let links = this.links;
        for (let len = links.length, i=0; i<len; i++) {
            if (links[i].name === name) {
                return links[i]
            }
        }
        console.log("Didn't find",name);
        return null;    // If not found
    }

    content(verbose) {
        console.assert(!this._needsfetch, "Must load before call content");
        if (this.data) { return this.data; }
        if (this.links) {
            let res = "";
            for (let i in this.links) {
                //noinspection JSUnfilteredForInLoop
                let l = this.links[i];
                res = res + l.content(verbose)  //Each link is a SB
            }
            return res;
        }
        console.log("ERR - object has no content, not even empty data");
        //Not supporting hash/fetch as async
        //(this.hash and Dweb.transport.rawfetch(hash = self.hash, verbose=verbose, **options)) or # Hash must point to raw data, not another SB
    }
    file() { console.assert(false, "XXX Undefined function StructuredBlock.file"); }
    size() { console.assert(false, "XXX Undefined function StructuredBlock.size"); }

    sign(commonlist, verbose) {
        /*
         Add a signature to a StructuredBlock
         Note if the SB has a _acl field it will be encrypted first, then the hash of the encrypted block used for signing.
         :param CommonList commonlist:   List its going on - has a ACL with a private key
         :return: sig so that CommonList can add to _list
         */
        if (!this._hash) this.p_store(verbose);  // Sets _hash immediately which is needed for signatures
        if (!commonlist._publichash) commonlist.p_store(verbose);    // Set _publichash immediately (required for Signature.sign)
        let sig = Signature.sign(commonlist, this._hash, verbose)
        this._signatures.push(sig);
        return sig;  // so that CommonList can add to _list
    }
    verify() { console.assert(false, "XXX Undefined function StructuredBlock.verify"); }


    earliestdate(){    // Set the _date field to the earliest date of any signature or null if not found
        if (!this._signatures) {
            this._date = null;
        } else {
            if (!this._date) {
                this._date = this._signatures[0]["date"];
                for (let i = 1; this._signatures.length > i; i++) {
                    if ( this._date > this._signatures[i]["date"]) {
                        this._date = this._signatures[i]["date"];
                    }
                }
            }
        }
        return this._date;
    }

    static compare(a, b) {
        if (a.earliestdate() > b.earliestdate()) { return 1; }
        if (b.earliestdate() > a.earliestdate()) { return -1; }
        return 0;
    }

    static test(transport, document, verbose) {
        if (verbose) console.log("StructuredBlock.test")
        return new Promise((resolve, reject) => {
            let el = document.getElementById("myList.0");
            console.log("XXX@make assert: el=", el);
            try {
                let teststr = "The well structured block"
                let sb = new StructuredBlock(null, {"data": teststr});
                let sb2;
                if (verbose) {
                    console.log("StructuredBlock.test sb=", sb);
                }
                sb.p_store(verbose)
                    .then(() => sb2 = new StructuredBlock(sb._hash, null))   // Creates, doesnt fetch
                    .then(() => sb2.p_fetch())
                    .then(() => {
                        if (verbose) console.assert(sb2.data === teststr, "SB should round trip", sb2.data, "!==", teststr)
                    })
                    /* //TODO-IPFS create a test set of SB's that have a path
                    .then(() => sb.p_path(["langs", "readme.md"], verbose, ["p_elem", el, verbose, null]))
                    //To debug, uncomment the el.textContent line in Transportable.p_elem
                     */
                    .then(() => {
                        if (verbose) console.log("StructuredBlock.test promises complete")
                        resolve({sb: sb, sb2: sb2});
                    })
                    .catch((err) => {
                        console.log("Error in StructuredBlock.test", err);   // Log since maybe "unhandled" if just throw
                        reject(err);
                    })
            } catch (err) {
                console.log("Caught exception in StructuredBlock.test", err)
                raise(err)
            }
        })
    }
}exports = module.exports = StructuredBlock;
